### ISSUE TITLE
Support for decimal logicaltype?

### ISSUE THREAD

**Issue by xgamer4** [marcosschroh/dataclasses-avroschema#84]
Hello,

I'm wondering if there's plans to support the [decimal logical type](http://avro.apache.org/docs/current/spec.html#Decimal) like there's support for the date/time logical types. If there's no plans, do you have suggested workarounds, or could you point me to the file(s) to modify so I could look into implementing it myself and putting up a pull request?

Thanks!

**Comment by marcosschroh** [comment]
Hi @xgamer4 

Yes, we want to support `decimal logical type`. Because is a special type that needs `precision` and `scale` seems that is not enough to use `decimal.Decimal`. So, I would create another `type`:

```python
# types.py

@dataclasses.dataclass
class Decimal(typing.Generic[T]):
    """
    Represents an Avro Decimal type

    precision (int): Specifying the number precision
    scale(int): Specifying the number scale. Default 0
    """
    precision: int
    scale: int = 0
    default: typing.Any = dataclasses.field(default=MissingSentinel)
    _dataclasses_custom_type: str = "Decimal" 

   # I am not sure if the following attributes are applicable, you can check it
    namespace: typing.Optional[str] = None
    aliases: typing.Optional[typing.List] = None

    def __repr__(self) -> str:
        return f"Decimal presition: {self.presision} scale:{self.scale}"
```

and then create your field:

```python
# fields.py

@dataclasses.dataclass
class DecimalField(BaseField):
    def get_avro_type(self) -> typing.Dict[str, typing.Any]:
        avro_type = {
            "type": DECIMAL,
            "name": self.get_singular_name(self.name),
            "presision": self.precision,
           "scale": self.scale
        }
        
        # check the following whether is necessary 
        if self.default.namespace is not None:
            avro_type["namespace"] = self.default.namespace

        if self.default.aliases is not None:
            avro_type["aliases"] = self.default.aliases
        ############

        return avro_type

    def get_default_value(self):
        ...

    def fake(self):
        ...
```

What do you think?

**Comment by xgamer4** [comment]
@marcosschroh So I still haven't figured out exactly how dataclasses-avroschema does its magic, so I'm not sure how helpful my input will be. 

For the code, I don't think Decimal would have a namespace - bytes is a primitive type so namespace isn't applicable if I understand the docs. Aliases should be valid though. 

For the first comment about Decimal, though, it looks like fastavro already supports Python `decimal.Decimal` to AVRO serialization and deserialization (at least according to their test coverage - I haven't actually tested it on my end). I think I see what you mean, though... if you were to just have a class property typed as a Decimal, dataclasses-avroschema wouldn't be able to extract the precision and scale information, and requiring only Decimal to be require an explicit `= field(...)` in the class declaration isn't clear or expected. So ok, yeah, I think I'm following now. Makes sense to me. 

https://github.com/fastavro/fastavro/blob/master/tests/test_logical_types.py starting at line 191 if you'd like the reference on fastavro tests.

**Comment by xgamer4** [comment]
@marcosschroh Took me a bit to wrap my head around exactly how everything works/find time to begin, but I've got a draft pull request up. It's still incomplete, but I wanted to get something there. 

That said, while working on it I found that the `decimal.Decimal` class does have enough information to build out an avro schema, so off that I have two ways to define a field:

```
Example(AvroModel):
    a_decimal: decimal.Decimal = decimal.Decimal('3.14')  # Define implicitly, scale = 2 and precision = 3
    sec_dec: decimal.Decimal = types.Decimal(scale=2, precision=3, [default=decimal.Decimal(...)])  # Set explicitly
```

I'd like your thoughts on the no-default case, though:
```
Example(AvroModel):
    no_default: decimal.Decimal
```

There is enough information to generate a schema off this, but what gets generated is non-obvious and likely confusing. Scale will be 0 (as the defined default from the Avro specs) and Precision just gets pulled from the default context - `decimal.Context().prec`. On my machine, this generates a decimal logical type with `scale = 0` and `precision = 28`, making the "decimal" field a glorified integer field, capable of storing integers up to 28-digits in length. That's how I currently have it written. I'm thinking about having it raise an exception if no default is provided instead, requiring that scale and precision get set either implicitly or explicitly. Thoughts?
